### PR TITLE
Reference objects in kernel

### LINKED ISSUE #29: Push down or inject assertions into kernel
We would want to be inject assertions (push down) into the kernel instead of a pull up approach. This would help in two ways:
- non JSON serializable objects can also be asserted against (as they cannot be pulled up in a straightforward manner)
- allow for kernel agnostic assertions (current implementation of pull up #26 only works with ipykernel)

Sample implementation by @MSeal 

```python
class NotebookTest:
    @context
    def exception_mapper():
        try:
            yield
        except CellExecutionError as e:
            # Extract exception and reraise closest exception type
            if (e.exception_class == 'AssertionError'):
                raise AssertionError(e.message)
    
    def assert_in_notebook(lhs_var, rhs):
        try:
            with self.exception_mapper():
                if (self.kernel_name == 'python'):
                    rhs_assign = papermill.translate('_assert_key', rhs, self.kernel_name)
                    code_to_inject = f"""
                    import pytest #???
                    {rhs_assign}
                    assert {lhs_var} == _assert_key
                    """
                else:
                    raise NotImplementedError("No json dump available for {kernel_name}")
                self.inject_code(code_to_inject)
        except AssertionError:
            # Might not be necessary
            print("Normal pytest message")
            
            
    class EmbededEqualityReference():
        def __init__(self, nbt, lhs, rhs):
            self.nbt = nbt
            self.lhs = lhs
            self.rhs = rhs
        
        def assert_in_notebook():
            self.nbt.assert_in_notebook(self.lhs, self.rhs)

            
    class EmbededVariableReference():
        # Custom collections.???
        # Compare with Mock objects
        def __init__(self, nbt, lhs):
            self.nbt = nbt
            self.lhs = lhs
            
        def __eq__(self, rhs):
            return EmbededEqualityReference(self.nbt, self.lhs, self.rhs)

        def resolve():
            return fetch_variable(self.lhs, self.nbt.kernel_name)


@pytest.hookimpl(hookwrapper=True)
def pytest_notebooktest_call(item):
    try:
        yield.get_result()
    except AssertionError as e:
        if (isinstance(e.args[0], EmbededEqualityReference):
            e.args[0].assert_in_notebook()
        # Continue with test call?
    except:
            raise
```

### PR DESCRIPTION
**Pull request by rohitsanj**
Resolves #29 

Implemented a method called `ref` which takes in the name of the object, and returns a testbook reference object.

The good thing about this is, it allows for normal unit tests to be written against functions/cells - which is the essence of testbook.

For example,

```python
from testbook import testbook
import pytest


@pytest.mark.parametrize("a, b, expected", [
    (
        (1, 0), (0, 1), 2 ** 0.5
    ),
    (
        (1, 0), (0, 0), 1
    ),
])
def test_calculate_distance(a, b, expected):
    with testbook('k-means.ipynb', prerun=['imports', 'kmeans2']) as tb:
        calculate_distance = tb.ref("calculate_distance")
        assert calculate_distance(a, b) == expected
```

However, we can only pass JSON serializable arguments to the calls -- to tackle this, we can provide a way to reference variables in the kernel without directly returning its value, something like `TestbookVariable` maybe, and then construct the code to be injected accordingly.

### CODE REVIEW

File: testbook/client.py [hunk 1]
```diff
@@ -9,10 +9,42 @@
...
+class TestbookCallable:
+    def __init__(self, tb, name):
+        self.tb: TestbookNotebookClient = tb
+        self.name: str = name
+
+    def __repr__(self):
+        return f"<TestbookCallable name='{self.name}'>"
+
+    def __call__(self, *args, **kwargs):
+        args_str = ', '.join(map(repr, args)) if args else ''
```
Comment (MSeal) [inline testbook/client.py:21]: Repr isn't the right thing to use here. I'd use json parsing or papermill's translator style conversion

File: testbook/client.py [hunk 2]
```diff
@@ -9,10 +9,42 @@
...
+        kwargs_str = (
```
Comment (MSeal) [inline testbook/client.py:22]: Same with kwargs, pass a json of the arguments and splat on other side.

File: testbook/client.py [hunk 3]
```diff
@@ -9,10 +9,42 @@
...
+            (', ' + ', '.join([f"{key} = {repr(value)}" for key, value in kwargs.items()]))
+            if kwargs
+            else ''
+        )
+        code = f"""
+            # Calling {self.name}
+            {self.name}({args_str + kwargs_str})
```
Comment (MSeal) [inline testbook/client.py:29]: {self.name}(*{args_list}, **{kwargs_dict})

File: testbook/client.py (hunk 1 above)
Comment (MSeal) [reply testbook/client.py:21]: e.g. `args_list = PythonTranslator.translate(args)`

File: testbook/translators.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
+import math
```
Comment (MSeal) [inline testbook/translators.py:1]: Docstring stating this was sourced from papermill

File: testbook/client.py [hunk 5]
```diff
@@ -146,16 +191,31 @@ def inject(self, code, args=None, run=False, before=None, after=None):
...
+            return cell.outputs[0].data['application/json']['value']
+        else:
+            return self._output_text(result)
+
+    @contextmanager
+    def exception_mapper(self):
+        try:
+            yield
+        except CellExecutionError as e:
+            raise eval(e.ename)(e) from None
```
[the remark this answers is not in the document]
Comment (rohitsanj) [reply testbook/client.py:212]: This is the output of `inspect.getmembers(e)`:
<details><summary>Output</summary>

```
[('__cause__', None),
 ('__class__', <class 'nbclient.exceptions.CellExecutionError'>),
 ('__context__', None),
 ('__delattr__',  <method-wrapper '__delattr__' of CellExecutionError object at 0x10d426a00>),
 ('__dict__',
  {'ename': 'NameError',
   'evalue': "name 'ref' is not defined",
   'traceback': '...'}),
 ('__dir__',  <built-in method __dir__ of CellExecutionError object at 0x10d426a00>),
 ('__doc__', '...'),
 ('__eq__',  <method-wrapper '__eq__' of CellExecutionError object at 0x10d426a00>),
 ('__format__',  <built-in method __format__ of CellExecutionError object at 0x10d426a00>),
 ('__ge__',  <method-wrapper '__ge__' of CellExecutionError object at 0x10d426a00>),
 ('__getattribute__',  <method-wrapper '__getattribute__' of CellExecutionError object at 0x10d426a00>),
 ('__gt__',  <method-wrapper '__gt__' of CellExecutionError object at 0x10d426a00>),
 ('__hash__',  <method-wrapper '__hash__' of CellExecutionError object at 0x10d426a00>),
 ('__init__',  <bound method CellExecutionError.__init__ of CellExecutionError("...")>),
 ('__init_subclass__',
  <built-in method __init_subclass__ of type object at 0x7ffa68825180>),
 ('__le__',  <method-wrapper '__le__' of CellExecutionError object at 0x10d426a00>),
 ('__lt__',  <method-wrapper '__lt__' of CellExecutionError object at 0x10d426a00>),
 ('__module__', 'nbclient.exceptions'),
 ('__ne__',  <method-wrapper '__ne__' of CellExecutionError object at 0x10d426a00>),
 ('__new__', <built-in method __new__ of type object at 0x10a864a68>),
 ('__reduce__',  <built-in method __reduce__ of CellExecutionError object at 0x10d426a00>),
 ('__reduce_ex__',  <built-in method __reduce_ex__ of CellExecutionError object at 0x10d426a00>),
 ('__repr__',  <method-wrapper '__repr__' of CellExecutionError object at 0x10d426a00>),
 ('__setattr__',  <method-wrapper '__setattr__' of CellExecutionError object at 0x10d426a00>),
 ('__setstate__',  <built-in method __setstate__ of CellExecutionError object at 0x10d426a00>),
 ('__sizeof__',  <built-in method __sizeof__ of CellExecutionError object at 0x10d426a00>),
 ('__str__',
  <bound method CellExecutionError.__str__ of CellExecutionError("...")>),
 ('__subclasshook__',  <built-in method __subclasshook__ of type object at 0x7ffa68825180>),
 ('__suppress_context__', False),
 ('__traceback__', <traceback object at 0x10d43f940>),
 ('__unicode__',  <bound method CellExecutionError.__unicode__ of CellExecutionError("..."),
 ('__weakref__', None),
 ('args', "...",
 ('ename', 'NameError'),
 ('evalue', "name 'ref' is not defined"),
 ('from_cell_and_msg', <bound method CellExecutionError.from_cell_and_msg of <class 'nbclient.exceptions.CellExecutionError'>>),
 ('traceback', '....'),
 ('with_traceback', <built-in method with_traceback of CellExecutionError object at 0x10d426a00>)]
```
</details>


Looks like we would need to make some changes to the way nbclient is raising the error since there is no `__bases__` attribute.

File: testbook/client.py [hunk 6]
```diff
@@ -150,18 +165,41 @@ def inject(self, code, args=None, run=True, before=None, after=None):
...
+
+        except ValueError:
+            varname = random_varname()
+            self.inject(
+                f"""
+                {varname} = _
+            """,
+                pop=True,
+            )
+            return TestbookVariableReference(self, varname)
```
Comment (rohitsanj) [inline testbook/client.py:197]: With this, we return a reference to the non-JSON serializeable object.

File: testbook/client.py [hunk 7]
```diff
@@ -7,12 +7,28 @@
 
 from .exceptions import TestbookCellTagNotFoundError, TestbookError
 from .testbooknode import TestbookNode
+from .translators import PythonTranslator
+from .reference import TestbookCallableReference, TestbookVariableReference
+from .utils import random_varname, _construct_call_code
+
+_testbook_object_cache = dict()
```
Comment (rohitsanj) [inline testbook/client.py:14]: I've implemented a caching system to store references, so that new reference objects don't get created for the same kernel variable/function.

Comment (rohitsanj) [reply testbook/client.py:14]: > Careful here, you just created a global cache object that's not scoped to a particular notebook. If you're going to cache make it an object on the TestbookNotebookClient instance.

Right! I didn't think of that! I will move it to the instance instead.

> Also if those references have state themselves this is doubly dangerous and you may not want to cache at all. 

I think this shouldn't be a problem because the reference object just holds a reference to the variable name in the kernel and not the value itself.. 

> What behavior was leading you to wanting a cache here?

To prevent redundant creation of multiple reference objects that reference the same variable/function in the kernel.

For example,
```python
@testbook('notebook.ipynb')
def foo(tb):
    assert tb.foo(1, 2) == 3 # new reference obj created
    assert tb.foo(2, 3) == 5 # yet another obj created
```

Comment (MSeal) [reply testbook/client.py:14]: You're not going to slow anything down by creating new reference objects here. I think we should ditch caching until there's an actual bottleneck. i.e. [premature-optimization root-of-evil](https://stackify.com/premature-optimization-evil/#:~:text=%E2%80%9CPremature%20optimization%20is%20the%20root,famous%20saying%20among%20software%20developers.&text=%E2%80%9CThe%20real%20problem%20is%20that,of%20it)%20in%20programming.%E2%80%9D)

File: testbook/client.py [hunk 8]
```diff
@@ -146,18 +165,44 @@ def inject(self, code, args=None, run=True, before=None, after=None):
...
+            outputs = self.inject(inject_code, pop=True).outputs
+            return outputs[0].data['application/json']['value']
+
+        except ValueError:
+            if not safe:
+                raise
+
+            varname = random_varname()
+            self.inject(
+                f"""
```
Comment (MSeal) [inline testbook/client.py:195]: This can just be `f"{varname = _}"`

File: testbook/client.py [hunk 9]
```diff
@@ -146,18 +165,44 @@ def inject(self, code, args=None, run=True, before=None, after=None):
...
+                {varname} = _
+            """,
+                pop=True,
+            )
+            return TestbookObjectReference(self, varname)
+
+    def _assert_in_notebook(self, lhs, rhs):
+        code = """
```
Comment (MSeal) [inline testbook/client.py:203]: Can also probably just do "assert {lhs} == {rhs}".format(...)

File: testbook/reference.py [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+        self.name: str = name
+
+    def __repr__(self):
+        return f"<TestbookObjectReference name='{self.name}' id={id(self)}>"
+
+    def __getattr__(self, name):
+        return TestbookObjectReference(self.tb, f"{self.name}.{name}")
+
+    def __eq__(self, rhs):
+        return self.tb._assert_in_notebook(self.name, rhs)
```
Comment (MSeal) [inline testbook/reference.py:13]: This should be a boolean, not an assertion

File: testbook/client.py [hunk 11]
```diff
@@ -146,18 +165,44 @@ def inject(self, code, args=None, run=True, before=None, after=None):
...
+            JSON({"value" : _})
         """
-        cell = self.inject(code)
-        return cell.outputs[0].data['application/json']['value']
+
+        try:
+            outputs = self.inject(inject_code, pop=True).outputs
+            return outputs[0].data['application/json']['value']
+
+        except ValueError:
```
Comment (MSeal) [inline testbook/client.py:189]: When we test let's make sure code error exceptions still raise:
e.g. injected code is invalid it shouldn't return a reference

File: testbook/client.py [hunk 12]
```diff
@@ -146,18 +165,44 @@ def inject(self, code, args=None, run=True, before=None, after=None):
...
+        inject_code = """
+            from IPython.display import JSON
+            JSON({"value" : _})
         """
-        cell = self.inject(code)
-        return cell.outputs[0].data['application/json']['value']
+
+        try:
+            outputs = self.inject(inject_code, pop=True).outputs
+            return outputs[0].data['application/json']['value']
```
Comment (MSeal) [inline testbook/client.py:187]: Let's make sure to test many different edge cases here

File: testbook/client.py [hunk 13]
```diff
@@ -146,18 +165,44 @@ def inject(self, code, args=None, run=True, before=None, after=None):
...
+
+        except ValueError:
+            if not safe:
```
Comment (MSeal) [inline testbook/client.py:190]: Let's try without `safe` first, and raise an exception saying maybe use `ref` method instead

File: testbook/client.py [hunk 14]
```diff
@@ -29,8 +50,6 @@ def _execute_result(cell):
 
     @staticmethod
     def _output_text(cell):
-        if not cell["outputs"]:
-            return
         text = ''
         for output in cell["outputs"]:
```
Comment (MSeal) [inline testbook/client.py:54]: This will blow up if you target a markdown cell because they don't have an output

File: testbook/client.py [hunk 15]
```diff
@@ -85,6 +104,10 @@ def execute_cell(self, cell, **kwargs):
 
         return executed_cells[0] if len(executed_cells) == 1 else executed_cells
 
+    def execute(self, **kwargs):
```
Comment (MSeal) [inline testbook/client.py:107]: why kwargs then?

File: testbook/client.py (hunk 13 above)
Comment (rohitsanj) [reply testbook/client.py:205]: I think we should have the option of returning a reference to the output if it cannot be serialized. 
I've made the default for `safe` as `False`.

The `value` method is more of an implementation/helper function, rather than a user-facing function, and we should encourage users to use the `ref` method to create references, and `inject` to inject executable code.

Comment (MSeal) [reply testbook/client.py:205]: I still think I'd rather make the user have to choose value or ref explicitly because the actions I can take after calling `value` are slightly different if I got a reference vs an actual value. But if you feel strongly to keeping it with the default we can try that out.

File: testbook/client.py (hunk 15 above)
Comment (rohitsanj) [reply testbook/client.py:107]: Whoops, I shall remove that.

File: testbook/client.py (hunk 13 above)
Comment (rohitsanj) [reply testbook/client.py]: Sure, one way we could do it is by adding an attribute to the error object containing the varname (for the non-JSON serializeable output), and then catch it elsewhere to create a reference to said varname.

Comment (MSeal) [reply testbook/client.py]: You could do that. Though I think in general you'll just change your code to call ref instead in that case.
